Change `??=` for nullable value types as specified in LDM

333fred opened the pull request:
Implement the changes from https://github.com/dotnet/csharplang/pull/2591. `a ??= b`, where `a` is `int?` and `b` is `int` now has a type of `int` instead of `int?`, and a different lowered form. @cston @dotnet/roslyn-compiler for review.

333fred commented:
Ping @dotnet/roslyn-compiler @cston for a review please.

jcouv commented:
Looking

jcouv commented on src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf:
```diff
@@ -1332,11 +1332,6 @@
-        <target state="translated">Hodnota default se převede na null, nikoli na default({0}).</target>
-        <note />
-      </trans-unit>
       <trans-unit id="WRN_DefaultLiteralConvertedToNullIsNotIntended_Title">
```
>WRN_DefaultLiteralConvertedToNullIsNotIntended_Title [](start = 22, length = 52)

`WRN_DefaultLiteralConvertedToNullIsNotIntended_Title` can also be removed

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -1158,13 +1481,12 @@ public static void Main()
-        // This should issue a warning, as we're converting to A, not A0
         int? i2 = null;
-        UseParam(i2 ??= default); // warn
+        UseParamAsInt(i2 ??= default);
```
>default [](start = 29, length = 7)

Please verify semantic model on `default`. Also, let's verify the resulting value in execution (could be done in a standalone test).
If I understood correctly, this is now `default(int)` (as opposed to `default(int?)`/`null` previously).

jcouv commented:
>             // must be A (where A is the type of a), as it is where the result of the assignment will end up. Therefore,

is this comment still correct?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3615 in 6977857. [](commit_id = 6977857c691552e98bf8ec23f629928ef84c182f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3623,17 +3623,29 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
+            if (leftType.IsNullableType())
             {
-                Error(diagnostics, ErrorCode.WRN_DefaultLiteralConvertedToNullIsNotIntended, node.Right, leftType.GetNullableUnderlyingType());
+                // If B is implicitly convertible to A0, then the result type of this expression is A0, except if B is dynamic.
```
>except if B is dynamic. [](start = 104, length = 23)

Why the exception for dynamic? Seems irregular

333fred replied:
This is explicitly called out in the spec, it's the same behavior that `??` has.

---
In reply to: [293040246](https://github.com/dotnet/roslyn/pull/36329#discussion_r293040246) [](ancestors = 293040246)

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3623,17 +3623,29 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
+                    diagnostics.Add(node, useSiteDiagnostics);
+                    var convertedRightOperand = CreateConversion(rightOperand, underlyingRightConversion, underlyingLeftType, diagnostics);
+                    // Create a nullable conversion for use by the local rewriter
+                    var nullableConversion = Conversions.ClassifyImplicitConversionFromType(underlyingLeftType, leftType, ref useSiteDiagnostics);
```
>nullableConversion [](start = 24, length = 18)

This conversion doesn't seem to be used (it's not saved into the bound node we create)

333fred replied:
Whoops, forgot to remove. I was using this at one point and then removed it from the bound node, but forgot to remove the creation.

---
In reply to: [293042697](https://github.com/dotnet/roslyn/pull/36329#discussion_r293042697) [](ancestors = 293042697)

333fred commented:
>             // must be A (where A is the type of a), as it is where the result of the assignment will end up. Therefore,

I've updated a few comments to be more explicit about types.

---
In reply to: [501382112](https://github.com/dotnet/roslyn/pull/36329#issuecomment-501382112) [](ancestors = 501382112)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3615 in 6977857. [](commit_id = 6977857c691552e98bf8ec23f629928ef84c182f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -2346,12 +2346,14 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
             LearnFromNullTest(leftOperand, ref this.State);
+            if (node.IsNullableValueTypeAssignment)
+            {
+                targetType = TypeWithAnnotations.Create(node.Type, NullableAnnotation.NotAnnotated);
```
>targetType = TypeWithAnnotations.Create(node.Type, NullableAnnotation.NotAnnotated); [](start = 16, length = 84)

Can we add a nullability test to cover this? (`(nullableInt ??= default).ToString()` or something along those lines)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                // if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
+                // tmp
+                //
+                // Except that a is only evaluated once.
```
>Except that a is only evaluated once. [](start = 19, length = 37)

I didn't understand this part. `a` does not appear in the above lowered form.

333fred replied:
Ignore, I'll remove this line.

---
In reply to: [293090135](https://github.com/dotnet/roslyn/pull/36329#discussion_r293090135) [](ancestors = 293090135)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                // Except that a is only evaluated once.
+
+                var leftOperand = node.LeftOperand;
+                if (!TryGetNullableMethod(leftOperand.Syntax,
```
>if (!TryGetNullableMethod(leftOperand.Syntax, [](start = 16, length = 45)

Do we have tests for those missing methods?

333fred replied:
See `CodeGenNullCoalescingAssignmentTests.MissingMembers`

---
In reply to: [293090659](https://github.com/dotnet/roslyn/pull/36329#discussion_r293090659) [](ancestors = 293090659)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+
+                // If lhsRead is a call, such as to a property accessor, save the result off to a temp. This doesn't affect
+                // the standard ??= case because it only uses lhsRead once.
+                if (lhsRead.Kind == BoundKind.Call)
```
>if (lhsRead.Kind == BoundKind.Call) [](start = 16, length = 35)

I didn't understand this. I would have expected `TransformCompoundAssignmentLHS` to  already have done what's needed, and we can just use `transformedLHS` that it returned.
FWIW, other scenarios that use `TransformCompoundAssignmentLHS` do not special-case calls.

333fred replied:
Correct, because other compound assignments do not have to read the value twice like we have to. It leaves calls to property accessors as is, so if we don't store the value to a temp we'll end up calling the property accessor twice.

---
In reply to: [293093678](https://github.com/dotnet/roslyn/pull/36329#discussion_r293093678) [](ancestors = 293093678)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                // We lower the expression to this form:
+                //
+                // var tmp = lhsRead.GetValueOrDefault();
+                // if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
```
>transformedAssignment [](start = 64, length = 21)

nit: should be `transformedLhs` for consistency with comments below

jcouv replied:
>!lhsRead.HasValue [](start = 23, length = 17)

Should this rather be a null test on `tmp`?

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                // !lhsRead.HasValue
+                var lhsReadHasValue = BoundCall.Synthesized(leftOperand.Syntax, lhsRead, hasValue);
+                var lhsReadHasValueNegation =
+                    MakeUnaryOperator(UnaryOperatorKind.BoolLogicalNegation, leftOperand.Syntax, method: null, lhsReadHasValue, lhsReadHasValue.Type);
```
>MakeUnaryOperator [](start = 20, length = 17)

Can this be `_factory.Not(...)`?

333fred replied:
I didn't know about these, thanks.

---
In reply to: [293096260](https://github.com/dotnet/roslyn/pull/36329#discussion_r293096260) [](ancestors = 293096260)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+
+                // { tmp = b; transformedLhs = tmp; }
+                var consequenceBuilder = ArrayBuilder<BoundStatement>.GetInstance(2);
+                consequenceBuilder.Add(new BoundExpressionStatement(node.Syntax, tmpAssignment) { WasCompilerGenerated = true });
```
>BoundExpressionStatement [](start = 43, length = 24)

Consider using `_factory.ExpressionStatement(...)`.

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                var consequenceBuilder = ArrayBuilder<BoundStatement>.GetInstance(2);
+                consequenceBuilder.Add(new BoundExpressionStatement(node.Syntax, tmpAssignment) { WasCompilerGenerated = true });
+                consequenceBuilder.Add(new BoundExpressionStatement(node.Syntax, transformedLhsAssignment) { WasCompilerGenerated = true });
+                var consequence = new BoundBlock(node.Syntax, locals: ImmutableArray<LocalSymbol>.Empty, consequenceBuilder.ToImmutableAndFree())
```
>BoundBlock [](start = 38, length = 10)

Consider using `_factor.Block(statement1, statement2)`

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                statementsBuilder.Add(ifHasValueStatement);
+
+                _needsSpilling = true;
+                return _factory.SpillSequence(temps.ToImmutableAndFree(), statementsBuilder.ToImmutableAndFree(), tmp);
```
>SpillSequence [](start = 32, length = 13)

Technically, we could do without the spilling, just using a `Sequence`. I don't know if it's worth it though.
```
// no statements, only expressions and side-effects in the following:
// var tmp = lhsRead.GetValueOrDefault();
// !lhsRead.HasValue ? { /*sequence*/ tmp = loweredRight; transformedAssignment = tmp; tmp } : tmp
```

333fred replied:
I think this version is likely more efficient IL wise, since it doesn't have the multiple `tmp`s at the end of a ternary. Also, it's the version that LDM approved, so I'd prefer to stick with the existing one.

---
In reply to: [293099162](https://github.com/dotnet/roslyn/pull/36329#discussion_r293099162) [](ancestors = 293099162)

jcouv replied:
I don't think LDM approved any specific IL-level lowered form. I recall we only discussed a C#-level lowered form and also guiding principles.
Don't forget that the lowered form here (described at the top of this local function) is not the final lowered form anyways, as the spilling will manipulate it further. Also, spilling makes this harder to reason about (exactly what is spilled? what is the final form?) and required some changes to `TransformCompoundAssignmentLHS` which I'm a bit concerned about.

I'm curious what Neal, Aleksey or Chuck think about this.

---
In reply to: [293103887](https://github.com/dotnet/roslyn/pull/36329#discussion_r293103887) [](ancestors = 293103887,293099162)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:
```diff
@@ -463,7 +463,7 @@ private BoundDynamicIndexerAccess TransformDynamicIndexerAccess(BoundDynamicInde
         /// The returned node needs to be lowered but its children are already lowered.
         /// </returns>
-        private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalLHS, ArrayBuilder<BoundExpression> stores, ArrayBuilder<LocalSymbol> temps, bool isDynamicAssignment)
+        private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalLHS, ArrayBuilder<BoundExpression> stores, ArrayBuilder<LocalSymbol> temps, bool isDynamicAssignment, bool needsSpilling = false)
```
>bool needsSpilling = false) [](start = 189, length = 27)

As commented in lowering of null-coalescing operator, I'm not sure we need spilling. Changes in this file feel unnecessary.

agocke commented on src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:
```diff
@@ -3623,17 +3623,29 @@ private BoundExpression BindNullCoalescingAssignmentOperator(AssignmentExpressio
+                    diagnostics.Add(node, useSiteDiagnostics);
+                    var convertedRightOperand = CreateConversion(rightOperand, underlyingRightConversion, underlyingLeftType, diagnostics);
+                    // Create a nullable conversion for use by the local rewriter
+                    var nullableConversion = Conversions.ClassifyImplicitConversionFromType(underlyingLeftType, leftType, ref useSiteDiagnostics);
```
Not sure I understand what's going on here. Are we throwing away the conversion? Don't we need to attach it to something for the LocalRewriter?

333fred replied:
See the comment to Julien. I was using this at one point and just forgot to remove it.

---
In reply to: [293105993](https://github.com/dotnet/roslyn/pull/36329#discussion_r293105993) [](ancestors = 293105993)

agocke commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs:
```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                // the standard ??= case because it only uses lhsRead once.
+                if (lhsRead.Kind == BoundKind.Call)
+                {
+                    var lhsTemp = _factory.StoreToTemp(lhsRead, out var store, kind: SynthesizedLocalKind.Spill);
```
Why do we need `SynthesizedLocalKind.Spill`?

333fred replied:
When we get to the SpillSequenceSpiller, it asserts that all synthesized locals are long lived. The default (`LoweringTemp`) is not.

---
In reply to: [293111279](https://github.com/dotnet/roslyn/pull/36329#discussion_r293111279) [](ancestors = 293111279)

agocke replied:
I see no good reason why we need this temporary to be long lived, since the other ones used in compound assignments are not.

333fred replied:
The other ones are not used in a spill sequence, this is.

---
In reply to: [293135135](https://github.com/dotnet/roslyn/pull/36329#discussion_r293135135) [](ancestors = 293135135)

agocke commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:
```diff
@@ -491,6 +491,7 @@ private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalL
             // we have a variable on the left. Transform it into:
             // ref temp = ref variable
             // temp = temp + value
+            var localKind = needsSpilling ? SynthesizedLocalKind.Spill : SynthesizedLocalKind.LoweringTemp;
```
Seems worrying if we create a ref temp that needs spilling since that's not possible.

333fred replied:
Turns out this is an existing problem with compound assignment in general. Filed [https://github.com/dotnet/roslyn/issues/36443](https://github.com/dotnet/roslyn/issues/36443)

---
In reply to: [293113591](https://github.com/dotnet/roslyn/pull/36329#discussion_r293113591) [](ancestors = 293113591)

jcouv commented:
>         /// <param name="isDynamicAssignment">True if the compound assignment is a dynamic operation.</param>

nit: Let's xml doc the new parameter: `needsSpilling` should be `true` when the variable produced by `TransformCompoundAssignmentLHS` for the LHS should have "spill" kind.

---
Refers to: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:461 in 6977857. [](commit_id = 6977857c691552e98bf8ec23f629928ef84c182f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:
```diff
@@ -560,16 +561,16 @@ private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalL
                             var loweredIndices = VisitList(arrayAccess.Indices);
 
-                            return SpillArrayElementAccess(loweredArray, loweredIndices, stores, temps);
+                            return SpillArrayElementAccess(loweredArray, loweredIndices, stores, temps, localKind);
```
>localKind); [](start = 104, length = 11)

I'm assuming that all the cases in this switch are covered by a nullable value type scenario. If not, let's add test coverage.

333fred replied:
Added array accesses. We had some for indexers, but not for regular arrays. Everything else should be covered already.

---
In reply to: [293121299](https://github.com/dotnet/roslyn/pull/36329#discussion_r293121299) [](ancestors = 293121299)

jcouv commented on src/Test/Utilities/Portable/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -813,6 +813,7 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
                                 ((isFirstOperandOfDynamicOrUserDefinedLogicalOperator(reference) ||
                                      isIncrementedNullableForToLoopControlVariable(reference) ||
                                      isConditionalAccessReceiver(reference) ||
+                                     isCoalesceAssignment(reference) ||
```
>isCoalesceAssignment [](start = 37, length = 20)

Could we tighten this exception to only be for null-coalescing assignments where the LHS is a property?

333fred replied:
I've updated the syntax used in the CFG to be a little more precise and removed this entirely.

---
In reply to: [293124296](https://github.com/dotnet/roslyn/pull/36329#discussion_r293124296) [](ancestors = 293124296)

333fred replied:
To be clear, I actually didn't need it last iteration, I had left it in but had apparently already removed the need.

---
In reply to: [294962493](https://github.com/dotnet/roslyn/pull/36329#discussion_r294962493) [](ancestors = 294962493,293124296)

jcouv commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ICoalesceAssignmentOperation.cs:
```diff
@@ -248,56 +248,53 @@ static void M()
-
 .locals {R1}
 {
-    CaptureIds: [0]
```
>[0] [](start = 16, length = 3)

Just curious: what previously caused only one capture to be reported here? (we already had `IFlowCaptureOperation: 1` below)

333fred replied:
The capture was more locally scoped to just Region 2.

---
In reply to: [293125795](https://github.com/dotnet/roslyn/pull/36329#discussion_r293125795) [](ancestors = 293125795)

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2874,49 +2874,129 @@ public override IOperation VisitCoalesceAssignment(ICoalesceAssignmentOperation
+                if (!isStatement)
+                {
+                    intermediateFrame = PushStackFrame();
+                    SpillEvalStack();
```
>SpillEvalStack [](start = 20, length = 14)

not related to this PR: I feel like I forgot what most of those utility methods do since last time I reviewed some CFG work. It may be worth investing in some documentation for CFG so that we can reload some context in future reviews.

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2874,49 +2874,129 @@ public override IOperation VisitCoalesceAssignment(ICoalesceAssignmentOperation
+
+            return isStatement ? null : GetCaptureReference(resultCaptureId, operation);
+
+            void nullableValueTypeReturn()
```
>nullableValueTypeReturn [](start = 17, length = 23)

```
// var tmp = lhsRead.GetValueOrDefault();
// if (!lhsRead.HasValue) { tmp = loweredRight; transformedAssignment = tmp; }
// tmp
// OR for statement
// if (!lhsRead.HasValue) {  transformedAssignment = loweredRight; }
```

jcouv commented:
>         Console.WriteLine(i1 ??= GetInt());

Consider adding a test for `(i1 ??= GetInt()).HasValue` which produces an error because the result type no longer is `int?`.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:29 in 6977857. [](commit_id = 6977857c691552e98bf8ec23f629928ef84c182f, deletion_comment = False)

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -145,67 +147,97 @@ public void FieldLvalue()
     {
         var c = new C();
         Console.WriteLine(c.f1 ??= GetInt());
+        Console.WriteLine(c.f2 ??= GetObject());
```
> Console.WriteLine(c.f2 ??= GetObject()); [](start = 7, length = 41)

Consider extracting both `WriteLine` statements to separate methods for clarity in `VerifyIL`.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:
```diff
@@ -338,85 +418,169 @@ public class C
+    public static void TestObject()
+    {
+        var c = new C();
+        Console.WriteLine(c.P2 ??= GetInt(1));
```
>1 [](start = 42, length = 1)

Consider incrementing here (use `3` and `4` instead of `1` and `2` again)

333fred commented:
>         Console.WriteLine(i1 ??= GetInt());

I don't think this is necessary. I already have several tests that show that type of this is `int`, not `int?`, and adding this case doesn't test anything new.

---
In reply to: [501481710](https://github.com/dotnet/roslyn/pull/36329#issuecomment-501481710) [](ancestors = 501481710)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs:29 in 6977857. [](commit_id = 6977857c691552e98bf8ec23f629928ef84c182f, deletion_comment = False)

333fred commented:
@agocke @jcouv I've addressed your feedback. I've moved to Julien's suggested lowering form for now, and filed [https://github.com/dotnet/roslyn/issues/36548](https://github.com/dotnet/roslyn/issues/36548) to follow up on it later.
